AdblockPlusSafari/RootController.m

```diff
@@ -48,4 +72,97 @@ - (void)setViewControllers:(NSArray<UIViewController *> *)viewControllers animat
   [super setViewControllers:viewControllers animated:animated];
 }
 
+- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSString *,id> *)change context:(void *)context
+{
+  if ([keyPath isEqualToString:NSStringFromSelector(@selector(activated))]
+      || [keyPath isEqualToString:NSStringFromSelector(@selector(needsDisplayErrorDialog))]) {
+    if (_readyToShowDialog) {
+      [self showDialogIfNeeded];
+    }
+  } else {
+    [super observeValueForKeyPath:keyPath ofObject:object change:change context:context];
+  }
+}
+
+- (void)setAdblockPlus:(AdblockPlusExtras *)adblockPlus
+{
+  NSArray<NSString *> *keyPaths = @[NSStringFromSelector(@selector(activated)),
+                                    NSStringFromSelector(@selector(needsDisplayErrorDialog))];
+
+  for (NSString *keyPath in keyPaths) {
+    [_adblockPlus removeObserver:self
+                      forKeyPath:keyPath];
+  }
+  _adblockPlus = adblockPlus;
+  for (NSString *keyPath in keyPaths) {
+    [_adblockPlus addObserver:self
+                   forKeyPath:keyPath
+                      options:NSKeyValueObservingOptionInitial|NSKeyValueObservingOptionNew
+                      context:nil];
+  }
+}
+
+#pragma mark - Private
+
+- (void)showDialogIfNeeded
+{
+  if ([DialogPresenterController shouldShowWelcomeDialogController:self.adblockPlus]) {
+    [self showWelcomeDialogIfNeeded];
+    return;
+  }
+
+  [self showErrorDialogIfNeeded];
+}
+
+- (void)showWelcomeDialogIfNeeded
+{
+  UIViewController *presenter = self;
+
+  while (presenter.presentedViewController) {
+    presenter = presenter.presentedViewController;
+    if ([presenter isKindOfClass:[DialogPresenterController class]]) {
+      return;
+    }
+  }
+
+  DialogPresenterController *controller = [self.storyboard instantiateViewControllerWithIdentifier:NSStringFromClass([DialogPresenterController class])];
+  controller.modalTransitionStyle = UIModalTransitionStyleCrossDissolve;
+  controller.modalPresentationStyle = UIModalPresentationOverCurrentContext;
+  controller.adblockPlus = self.adblockPlus;
+  [presenter presentViewController:controller animated:YES completion:nil];
+}
+
+- (void)showErrorDialogIfNeeded
+{
+  if (!self.adblockPlus.needsDisplayErrorDialog) {
+    return;
+  }
+
+  // Do not show message, if update was automatic.
+  if ([[self.adblockPlus.filterLists.allValues valueForKeyPath:@"@sum.userTriggered"] integerValue] == 0) {
+    return;
+  }
+
+  UIViewController *viewController = UIApplication.sharedApplication.delegate.window.rootViewController;
+
+  while (viewController.presentedViewController) {
+    viewController = viewController.presentedViewController;
+    if ([viewController conformsToProtocol:@protocol(DialogControllerProtocol) ]) {
+      return;
+    }
+  }
+
+  NSString *title = NSLocalizedString(@"​Filter list update failed",
```

Comment:
Is this already in Localizable.strings ? I see it was in previous revision but i have lost the context of what previous state the revision represents.